HIP/hip-0000-introduction-of-blocks.md

```diff
@@ -0,0 +1,199 @@
+
+**Historical State**
+
+`Record files` prior to the introduction of the block properties will not expose the required information. In order to be able to answer historical queries, `mirror nodes` will need to iterate over the existing/already imported record files since genesis, assign the corresponding `block number`, `hash` and migrate their DB collections such that they are able to answer queries of type:
+
+- Getting block by hash/number
+- Getting a list of transactions for a block by number/hash
+- Getting logs based on block number filtering
+
+### Block Properties
+
+The following table specifies all `block` properties and at which point they will be computed. Mirror nodes must `hex` encode all properties prior to exposing them through their APIs. This table defines the properties that must be returned through APIs from Mirror Nodes.
+
+| Property | Computed By | Description  |
+| --- | --- | --- |
+| number | Services | Stored in services state (only the current number) and exported in the record stream. It is the consecutive number of the record file that is being incremented by 1 for every new record file. The number will be initially set in services through the bootstrapping process. It is exposed to the 1) EVM during Transaction Execution through the NUMBER opcode; 2) as a new property in the record file and ingested by mirror nodes. |
+| timestamp | Services | Stored in services state (only the last block timestamp) and computed by services. It is the consensusTimestamp of the first transaction in the record file. It is exposed to the 1) EVM during Transaction Execution through the TIMESTAMP opcode; 2) implicitly exported in the record file through the TS of the first transaction in the record file. |
+| hash | Services | Stored in services (last 256 blocks). It is the keccak256(runningHash) of the previous record file. That is the running hash of the last record stream object from the previous record file. It is exposed to the 1) EVM during Transaction Execution through the BLOCKHASH opcode; 2) In the record file as the End Object Running Hash |
+| baseFeePerGas (??) | Mirror Node | Computed by Mirror Node(s). It is a hex encoded value of the gas pricing based in the FeeSchedule file. |
+  | difficulty | Mirror Node | Hardcoded by Mirror Node(s) to hex encoded 0 |
+  | extraData | Mirror Node | Hardcoded by Mirror Nodes(s) to 0x |
+  | gasLimit | Mirror Node |  Computed by Mirror Node(s). The sum of the gasLimit values for all ContractLocalCall, ContractCall and ContractCreate transactions within the block. |
```

Comment:
`ContractLocalCall` won't appear in the record stream as it does not contribute to consensus. Same comment for other references to it.

Reply:
This should be some function of the block size and gas throttle, as it represents the theoretical max not the actual.
```suggestion
  | gasLimit | Mirror Node |  Computed by Mirror Node(s). The gas throttle limit per second multiplied by the target block time. |
```

Reply:
Do mirror nodes have the necessary information to compute that? @tinker-michaelj do we export throttle configurations in any way currently and if we do, @steven-sheehy do you know whether they are ingested by mirror nodes?

Reply:
The gas throttles are available in network file 0.0.121 app.properties, but only if overridden. So in practice, no, they're not available to mirror node since they're not usually overridden.

Reply:
I've updated the HIP with an Open Question related to the initial "feed" of the values.